Importing robot from URDF/XACRO file (#2378)
o3de/o3de.org: content/docs/user-guide/interactivity/robotics/importing-robot.md

#### Loading mesh files

6. If the XACRO or URDF file is loaded correctly you will see a table with a list of all mesh files (assets) used in the model. Each row of the table describes one asset, with information divided into the following columns:

- **URDF mesh path** - path extracted from the URDF
- **Resolved mesh from URDF** - file path of the mesh located in the filesystem
- **Product asset** - location of the asset description (`*.azmodel` or `*.pxmesh`) in the O3DE project assets folder
- **Source asset** - location of the asset mesh file in the O3DE project assets folder
- **Type** - the type of the asset, which can be `visual` or `collider`

Besides this information, there is an indication if the asset was imported correctly (green) or if the import failed (red).

Additional information about each asset can be obtained by double-clicking its row. This will open the O3DE Asset Processor showing import details. This information may be especially helpful in case of assets that were not imported correctly.

Click **Next** when finished.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_mesh_list.png)

**What to do if there are problems with importing assets?**

> 

Reviewer: ```suggestion
If the Source asset field is marked as failed, it is caused by a URDF importer failure to find a reference to the mesh file in the file system. In such a situation, double-check whether the robot's description package was built and sourced correctly.
If the Product asset field is marked as failed, it is caused by the Asset Builder failing to build the product asset. Refer to the Asset Processor output to resolve the issue.
```

 Thus typically robot models are made available in packages with additional files containing visualizations and collision shapes. Typical formats of mesh files in robotics are [Collada DAE](https://en.wikipedia.org/wiki/COLLADA) and [STL](https://en.wikipedia.org/wiki/STL_(file_format)). It is also a common practice to make available such robot packages as ROS workspaces, which make them easier to use in ROS applications.

### URDF/XACRO importer - general description

Both URDF and XACRO files can be imported into the O3DE using the importer included in the ROS2 Gem. The main features of the importer:

- The importer can read both: URDF, as well as XACRO files.
- The importer takes into account all features of the XACRO format, such as includes and variables.
- Mesh files can be copied automatically to the asset folder of the O3DE project.
- The importer creates a multi-body structure using articulations or classic rigid bodies and joints components.

### Importing URDF/XACRO - step-by-step

> > Before you start:

Reviewer 2: ```suggestion
```

> 1. Importing URDF/XACRO files requires ROS-ready project with a ROS Gem enabled. Please refer to the [Project Configuration section](project-configuration.md) for details.
> 2. In this section it is assumed, you have already created a O3DE project. Additional information on creating a robotic project can be found in the [Creating a Robotic Simulation section](creating-robotic-simulation.md)

> > 3. When using a XACRO files which are a part of a ROS workspace, it may be required to build the workspace first. Otherwise, the importer may fail to find all required files. Please refer to the documentation of the robot's description package for building instructions (typically you have to place it in the `src` folder of your ROS workspace and perform `colcon build`).

Reviewer 2: ```suggestion
When using robot definition files that are a part of a ROS package, you will often need to build a workspace with such package and its dependencies first. Otherwise, the importer may fail to find all required files. Please refer to the documentation of your robot description package, which is often included in its GitHub README file. Building typically involves creating a workspace and buidling it with `colcon build`.
```

> #### Parsing URDF/XACRO file

Reviewer 2: ```suggestion
### Loading the robot definition file with Robot Importer
```

3. Click **Next** on the information page

4. On the next page:

- Select a URDF or XACRO file to be imported by clicking on the **[...]** button
- Select (using the checkbox) if the importer should copy meshes into the O3DE project assets folder or not
- Click **Next**

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_load_file.png)

> Note: if you leave the **Import meshes...** checkbox unchecked, the importer will try to find matching meshes in the project asset library. Verification of identity of meshes is done by CRC comparison.

5. If you are importing a XACRO file that features parameters to be set, you will be asked to do so on the next page. Parameters can be edited by double-clicking on the value. Click **Next** when finished.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_XACRO_parameters.png)

> Note: which parameters must be set for succesful import strictly depends on the XACRO project you are importing. Please refer to its documentation for more information.

If import fails you will see a message with the output of the ROS 2 XACRO executable.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_fail.png)

> #### Loading mesh files

Reviewer 2: ```suggestion
### Processing robot assets
```

Besides this information, there is an indication if the asset was imported correctly (green) or if the import failed (red).

Additional information about each asset can be obtained by double-clicking its row. This will open the O3DE Asset Processor showing import details. This information may be especially helpful in case of assets that were not imported correctly.

Click **Next** when finished.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_mesh_list.png)

**What to do if there are problems with importing assets?**
If the Source asset field is marked as failed, it is caused by a URDF importer failure to find a reference to the mesh file in the file system. In such a situation, double-check whether the robot's description package was built and sourced correctly.
If the Product asset field is marked as failed, it is caused by the Asset Builder failing to build the product asset. Refer to the Asset Processor output to resolve the issue.
You can do either:

- click **Back**, resolve problems with mesh files and click **Next** to retry import, or
- proceed by clicking **Next** and finish URDF import. Missing meshes can be added afterward in the O3DE editor.

> #### Creating O3DE prefab

Reviewer 2: ```suggestion
### Creating a robot prefab
```

7. On the next page:

- Choose a prefab name

> - Indicate (using the checkbox) if robot physics should be based on articulations, or classic rigid bodies and joints structure.

Reviewer 2: ```suggestion
- Indicate (using the checkbox) if robot physics should be based on articulations or classic rigid bodies and joints structure.
```

Click **Create Prefab** when finished.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_prefab_creation.png)

> > Note: the decision whether to use articulations highly depends on the robot structure and simulation requirements. In general, articulations are more stable and computational efficient, but have some limitations. The most important limtation is that articulations natively only support tree-structures - closed loops are not allowed. The "rule of thumb" would be to use articulations in robotic simulation whenever it is possible. However please refer to [this documentation](https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/Articulations.html) for more details.

Reviewer 2: This part of the documentation should be adapted to the fact that the release will have PhysX 4 not 5 by default. For this version, O3DE does not support articulations.

8. The last page of the wizard shows a summary of entities.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_summary.png)

> ### Additional information

Reviewer 2: ```suggestion
### Re-importing robots
```

URDF/XACRO files contain complete robot descriptions, including references to external geometry files. While primitive geometries can be defined directly within the URDF/XACRO file, it is common practice to use external mesh files in formats such as DAE (Collada) or STL to represent the visual and collision shapes of the robot.
Robot models are typically made available in packages that include the URDF/XACRO file and additional files containing visualizations and collision shapes, whether those shapes are primitive geometries or external mesh files. These packages may be distributed as ROS workspaces, which makes them easier to use in ROS applications. 

## Introduction to Robot Importer

Robots can be imported into O3DE simulation project with the importer tool included in the ROS 2 Gem. It has the following features:

- Guides you through the import process step by step.
- Reads URDF and XACRO files.
- It copies all required mesh files to the asset folder of your O3DE project.
- It creates a prefab with a multi-body structure using articulations or classic rigid bodies and joints components.

## Importing a robot into your simulation project

> When using robot definition files that are a part of a ROS package, you will often need to build a workspace with such package and its dependencies first. Otherwise, the importer may fail to find all required files. Please refer to the documentation of your robot description package, which is often included in its GitHub README file. Building typically involves creating a workspace and buidling it with `colcon build`.

Reviewer 2: This is an excellent suggestion, thank you!

### Loading the robot definition file with Robot Importer

1. Load your project and select a level.

2. Launch URDF/XACRO import wizard. This can be done by selecting **Main Menu > Tools > Robot Importer** or by clicking the **Robot Importer** icon in the toolbar.
  
![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_button.png)

3. Click **Next** on the information page

4. On the next page:

- Select a URDF or XACRO file to be imported by clicking on the **[...]** button
- Select (using the checkbox) if the importer should copy meshes into the O3DE project assets folder or not
- Click **Next**

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_load_file.png)

> > Note: if you leave the **Import meshes...** checkbox unchecked, the importer will try to find matching meshes in the project asset library. Verification of identity of meshes is done by CRC comparison.

Reviewer 2: a clear improvement, thank you!

- **Source asset** - location of the asset mesh file in the O3DE project assets directory
- **Type** - the type of the asset, which can be `visual` or `collider`

Besides this information, there is an indication if the asset was imported correctly (green) or if the import failed (red).

Additional information about each asset can be obtained by double-clicking its row. This will open the O3DE Asset Processor showing import details. This information may be especially helpful in case of assets that were not imported correctly.

Click **Next** when finished.

![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_mesh_list.png)

**What to do if there are problems with importing assets?**
If the Source asset field is marked as failed, it is caused by a URDF importer failure to find a reference to the mesh file in the file system. In such a situation, double-check whether the robot's description package was built and sourced correctly.
If the Product asset field is marked as failed, it is caused by the Asset Builder failing to build the product asset. Refer to the Asset Processor output to resolve the issue.

> You can do either:
> 
> - click **Back**, resolve problems with mesh files and click **Next** to retry import, or
> - proceed by clicking **Next** and finish URDF import. Missing meshes can be added afterward in the O3DE editor.

Reviewer 3: Add newline to create separate paragraph
Clarify the instructions you're providing and what they do. 
```suggestion

You can either resolve the issues now or after you finish using Robot Importer: 
- To resolve, click **Back** and resolve problems with the mesh files. Then click **Next** to retry import.
- To resolve later, click **Next** and finish URDF import. Then add the missing meshes in the O3DE Editor.
```

### Creating a robot prefab

> 7. On the next page:
> 
> - Choose a prefab name
> - Indicate (using the checkbox) if robot physics should be based on articulations or classic rigid bodies and joints structure.
> 
> Click **Create Prefab** when finished.
> 
> ![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_prefab_creation.png)

Reviewer 3: ```suggestion
1. Create a prefab by doing the following:

   - Enter a prefab name.
   - (Optional) Enable the checkbox if you want to base the robot physics on articulations or classic rigid bodies and joints structure.
    - Click **Create Prefab**.

    ![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_prefab_creation.png)
```

> > Note: the decision whether to use articulations highly depends on the robot structure and simulation requirements. In general, articulations are more stable and computational efficient, but have some limitations. The 2 main limtations are:
> >
> >- Articulations are only available in PhysX 5. By default O3DE is build with PhysX 4.
> >- Articulations natively only support tree-structures - closed loops are not allowed.
> >
> >The "rule of thumb" would be to use articulations in robotic simulation whenever it is possible. However please refer to [this documentation](https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/Articulations.html) for more details.

Reviewer 3: - Move to note callout box
- When linking to docs, use the name or description of the docs. If external docs, you must specify the source (who owns the docs)
```suggestion
    {{< note >}}
**Use articulation for joints and rigid bodies** depends on the robot structure and simulation requirements. In general, articulations are more stable and computational efficient, but have some limitations. The 2 main limtations are:
- Articulations are only available in PhysX 5. By default, O3DE is built with PhysX 4.
- Articulations only support tree-structures natively. Closed loops are not allowed.

We recommend to use articulations in robotic simulation whenever possible. For more details, refer to [Articulations](https://nvidia-omniverse.github.io/PhysX/physx/5.1.3/docs/Articulations.html) in NVIDIA PhysX 5 documentation.
    {{< /note >}}
```

> 8. The last page of the wizard shows a summary of entities.

Reviewer 3: ```suggestion
1. After creating a prefab, review a summary of the generated entities.
```

> ![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_summary.png)

Reviewer 3: ```suggestion
    ![Robot Importer](/images/user-guide/gems/ros2/URDF_importer_summary.png)
```

### Re-importing robots

> - The URDF importer can be used to import URDF files that were already imported. In such cases, however, assets (mesh files) are not updated if they were previously imported correctly. If you intend to perform re-import of assets, you should delete them from O3DE project. This can be done by using the Asset Browser.

Reviewer 3: Use consistent naming, "Robot Importer"
```suggestion
Use the Robot Importer to re-import URDF files. In some cases, assets (mesh files) don't update if they were previously imported correctly. To work around this and do a complete re-import, delete the assets from your project by using the Asset Browser or in the directory.
```
